ENH: add geometric_slerp

tylerjereddy opened the pull request:
**Purpose**: Provide an N-dimensional implementation of [geometric spherical linear interpolation](https://en.wikipedia.org/wiki/Slerp#Geometric_Slerp), which is not confined to 3-dimensions like [`scipy.spatial.transform.Slerp`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.transform.Slerp.html#scipy.spatial.transform.Slerp) & also not chained to infrastructure related to or understanding of quaternions.

**Context**: This is broken off from a much larger effort to support the calculation of the surface areas of spherical polygons in #8069. There, Slerp served the purpose of discretizing spherical surface polygons to achieve higher or lower accuracy in the surface area (integral) values, as desired. 

Arguably, having two different kinds of Slerp in SciPy may be contentious. I did try to [bring this up](https://github.com/scipy/scipy/pull/8994#discussion_r200010310), but didn't get a response when reviewing the original PR that uses quaternions. Regardless, I can always just move this to another repo if we don't want it in SciPy.

A more common / popular application of Slerp is for spherical trajectory simulation / animation on curved surfaces -- see my [example in matplotlib issue tracker from a while ago](https://github.com/matplotlib/matplotlib/issues/5294#issuecomment-367877945).

If we do like this, probably needs at least:

- [x] `asv` benchmarks
- [x] reduce verbosity in some places maybe

rgommers commented:
That still seems like a fairly small and digestible implementation, so I'm not immediately worried about maintenance. 

This is not exactly my area of expertise, let's see what @nmayorov thinks.

WarrenWeckesser commented:
Adding this function is OK with me.

I have some requests for changes:

* Shorten the argument name `t_values` to `t`. (The `values` suffix doesn't really add useful information to the name.)

* Accept array-like, so we can write, for example,

      geometric_slerp([1, 0], [0, 1], [0, 0.25, 0.5, 0.75, 1]) 

* t should accept an empty array.  The return value should also be empty in that case.  E.g.
    
      geometric_slerp([1, 0], [0, 1], []) 

  should return an array with shape (0, 2).

* It should accept a scalar for t.  E.g. this should return an array  with shape (2,):

      geometric_slerp([1, 0], [0, 1], 0.5)

* The handling of the 0-sphere is not correct.  The function accepts input values that are not on the 0-sphere, and returns values that are not on the 0-sphere:

      In [30]: p0 = np.array([1])

      In [31]: p1 = np.array([0])

      In [32]: t = np.array([0, 0.25, 0.5, 0.75, 1])

      In [33]: geometric_slerp(p0, p1, t)
      Out[33]: 
      array([[1.        ],
             [0.92387953],
             [0.70710678],
             [0.38268343],
             [0.        ]])

  The 0-sphere embedded in R¹ is a discrete set containing just two points, 1 and -1.  It doesn't make sense to interpolate between them.

Also, I don't think we need cython here.  It looks like the function could be implemented in a few lines of Python + NumPy.  Here's a straightforward translation of the formula from the wikipedia page (but with no argument validation):

    def slerp1(p0, p1, t):
        # This function assumes p0, p1 and t are 1-d.

        p0 = np.asarray(p0)
        p1 = np.asarray(p1)
        t = np.asarray(t)

        omega = np.arccos(np.dot(p0, p1))
        t = t.reshape(-1, 1)
        z = (np.sin((1 - t)*omega)*p0 + np.sin(t*omega)*p1) / np.sin(omega)

        return z

But that version doesn't handle a scalar `t` the way I suggested above.  Here's a version that allows `t` to have arbitrary dimension:

    def slerp2(p0, p1, t):
        # This version assumes p0 and p1 are 1-d, but
        # t can be n-d.  The return value has shape
        #    t.shape + (len(p0),).

        p0 = np.asarray(p0)
        p1 = np.asarray(p1)
        t = np.asarray(t)

        omega = np.arccos(np.dot(p0, p1))
        t = np.asarray(t)
        shp = t.shape
        t = t.reshape(-1, 1)
        z = (np.sin((1 - t)*omega)*p0 + np.sin(t*omega)*p1) / np.sin(omega)

        return z.reshape(shp + (len(p0),))

For example,

```
In [87]: slerp2([1, 0], [0, 1], 0.5)                                                                                         
Out[87]: array([0.70710678, 0.70710678])

In [88]: slerp2([1, 0], [0, 1], [0, 0.25, 0.5])                                                                              
Out[88]: 
array([[1.        , 0.        ],
       [0.92387953, 0.38268343],
       [0.70710678, 0.70710678]])

In [89]: slerp2([1, 0], [0, 1], [[0, 0.25], [0.5, 1]])                                                                       
Out[89]: 
array([[[1.        , 0.        ],
        [0.92387953, 0.38268343]],

       [[0.70710678, 0.70710678],
        [0.        , 1.        ]]])
```

-----

I also have a request for an enhancement, but I don't have a clear idea for an API.  Currently the function generates points along the shorter segment of the great circle.  How could we allow interpolation along the longer segment?

The implementation appears to be as simple as changing

    omega = np.arccos(np.dot(p0, p1))

to

    omega = 2*np.pi - np.arccos(np.dot(p0, p1))

I haven't worked through the math to double-check this, but I  tried it with the 1-sphere embedded in R², and it correctly gives points along the longer segment.

WarrenWeckesser commented:
A couple more issues:

* (I should have made this comment above, when I commented on `t_values`.)  The names `start_coord` and `end_coord` should be shortened to `start` and `end`.

* When the first two arguments are the same, the function returns `nan`s:

      In [142]: p0 = np.array([1, 0])                                                                                              

      In [143]: t = np.array([0, 0.3, 0.5, 0.9, 1])                                                                                

      In [144]: geometric_slerp(p0, p0, t)                                                                                         
      Out[144]: 
      array([[nan, nan],
             [nan, nan],
             [nan, nan],
             [nan, nan],
             [nan, nan]])

  (My Python versions of the function do this too.)

tylerjereddy commented:
Looks like a thorough review. I think I had a good reason for choosing Cython--if I ever get to the point where the spherical polygon surface area implementation happens, it is helpful to be able to call a Cython version for discretization with low overhead if there are millions of spherical polygons.

tylerjereddy commented:
I've pushed in a single revision commit that addresses the vast majority of Warren's review comments, with bullet points in the commit message outlining what was handled & how.

Two points on which we continue to disagree, but perhaps some sort of compromise might be reached on:

1) I maintain the value of the Cython implementation for building up computational geometry algorithms that loop through large numbers of spherical polygons & need to make frequent calls to `geometric_slerp()`

2) For this enhancement query:

> How could we allow interpolation along the longer segment?

I think it is fair to start off using the more common / geodesic use case. As the [article notes](https://en.wikipedia.org/wiki/Slerp#Geometric_Slerp):

> A Slerp path is, in fact, the spherical geometry equivalent of a path along a line segment in the plane; a great circle is a spherical geodesic.

Indeed, in a plane, that would be like asking for the longer distance between two points, which is the infinite radius of the special sphere that is the plane. While this could have some useful graphical applications on the sphere since it is finite there, I'd prefer to hold off on providing that use case until it is carefully considered.

tylerjereddy commented:
@pmla / @rtshort perhaps you two can take a look here as well?

nmayorov commented on scipy/spatial/_slerp.pyx:
```diff
@@ -0,0 +1,29 @@
+import numpy as np
```
Honestly, I'm not a fan of preliminarily optimization and "pre-over-engineering", but if you have a plan to use this code from Cython, then let this be.

nmayorov commented on scipy/spatial/_geometric_slerp.py:
```diff
@@ -0,0 +1,206 @@
+from . import _slerp
+from scipy.spatial.distance import euclidean
+
+def geometric_slerp(start,
```
Don't you think it can be useful to provide a class implementation, similar to, say, `PchipInterpolator` and `Slerp`?

The current function does interpolation along a single segment. Both can be provided.

tylerjereddy replied:
I'm not clear on the advantage of an object-oriented approach for this simple function? What is the use case, exactly?

nmayorov replied:
My suggestion was about making SLERP across many segments like quaternion SLERP [here](http://scipy.github.io/devdocs/generated/scipy.spatial.transform.Slerp.html#scipy.spatial.transform.Slerp).

This should reuse the function you implemented here and can be done separately. So let's don't consider it here.

nmayorov commented on scipy/spatial/_geometric_slerp.py:
```diff
@@ -0,0 +1,206 @@
+                    t,
+                    tol=1e-7):
+    """
+    Spherical linear interpolation along unit-radius
```
Nitpick --- the recommended formatting is one line followed by a more thorough explanation.

nmayorov commented on scipy/spatial/_geometric_slerp.py:
```diff
@@ -0,0 +1,206 @@
+        0 and 1. A common approach is to generate the array with
+        ``np.linspace(0, 1, n_pts)`` for linearly spaced points.
+        Ascending, descending, and scrambled orders are permitted.
+    tol: float
```
From the first look I'm not convinced that this parameter and the corresponding checking is necessary for this function. Seems not very important.

tylerjereddy replied:
I don't get it--how are your proposing that antipodal points are handled? That's a case of degenerate input, and allowing the user to control the sensitivity in detecting that degeneracy seems sensible to me.

nmayorov replied:
I see now that for antipodes division by zero may occur, so some guards are necessary. My bad.

However I see that the code posted by @pmla does not contain such division and according to his benchmarks work more stable for angles closes to `pi`. (I don't understand how it works though.)

I'd say that if we can avoid numerical instability then we should definitely do so. And in this case I would advocate for removing `tol`. Even if we have the situation with truly ambiguous interpolation curve --- it is still a valid curve and I believe a user should decide separately whether he wants such situation or not.

tylerjereddy replied:
Although I've removed the Cython code & used Peter's implementation, there were stability tradeoffs elsewhere (reproducing values exactly at 0) noted in my commit messages. 

I'm still maintaining that the infinite possible ambiguous paths between antipodes should raise rather than potentially returning one of those possible values.

pmla replied:
I had a look into the stability issues. It turns out that they result from the tiny errors in `np.sin` and `np.cos`. The previous implementation used only `np.sin`, whereas the current one uses `np.cos` and `np.sin`. Since one of the test inputs is `t=1/3` we have:
```
>>> np.cos(np.pi / 3) - 0.5
1.1102230246251565e-16
>>> np.sin(np.pi / 3) - np.sqrt(3) / 2
0
```
The use of `np.cos` means we cannot reproduce 0 exactly.

However, if a test input were `t=1/4` we would have:
```
>>> np.cos(np.pi / 4) - np.sqrt(2) / 2
0.0
>>> np.sin(np.pi / 4) - np.sqrt(2) / 2
-1.1102230246251565e-16
```
Now the use of `np.sin` means we cannot reproduce 0 exactly.

In other words, the ability of the previous implementation to reproduce values at 0 depends on the choice of input.

tylerjereddy replied:
The stability was not clearly superior in all aspects for the new implementation, but I can live with that tradeoff if it makes reviewers happy.

The antipodal discussion is happening in two places now, but the plot/analysis below should be clear on why I'm pushing for an exception over a warning once a user is close to antipodal input. We could push the threshold to the absolute limit, but if the algorithm changes under the hood the pathological threshold may change anyway---if you're getting close to antipodes you're probably doing something wrong and should consider just using the function twice & splitting up your workflow to get normal paths that may be fused.

nmayorov commented on scipy/spatial/_geometric_slerp.py:
```diff
@@ -0,0 +1,206 @@
+        object.
+        End coordinate for interpolation will be converted to
+        float64.
+    t: array
```
Might add the shape description as `float or (n_points,) array-like`.

And `(D, )` I would maybe change to `(n_dimensions,)`

nmayorov commented on scipy/spatial/_geometric_slerp.py:
```diff
@@ -0,0 +1,206 @@
+    end = np.asanyarray(end)
+
+    if start.ndim != 1 or end.ndim != 1:
+        raise ValueError("Start and end coordinates must be flat")
```
must be one-dimensional?

nmayorov commented on scipy/spatial/_geometric_slerp.py:
```diff
@@ -0,0 +1,206 @@
+    Parameters
+    ----------
+    start : (D, ) array-like
+        Single D-dimensional input coordinate in a 1-D array-like
```
Mention that `D` must be greater than 1.

Mention that `start` and `end` must lay on a unit sphere (have unit norm). Or as I've written below --- it can be replaced by normalization (or optional normalization).

nmayorov commented on scipy/spatial/_geometric_slerp.py:
```diff
@@ -0,0 +1,206 @@
+    # similarly for points that violate equation for n-sphere
+    # NOTE: this is tricky to think about for the 0-sphere;
+    # performing the check for 1-sphere (circle) and up
+    if start.size > 1:
```
From my perspective this check must go first. And I suggest that you should use `np.linalg.norm` for simplicity.

As an alternative, we can do the normalization here instead of checking.

tylerjereddy replied:
Why should this check go first? The transplant causes 7 tests failures, and I'm not clear that this violation is definitely more pathological than some of the others that precede it. 

Unless we have useful information on the average input this function is likely to receive over time, it seems to be somewhat speculative to reorder the error handling.

I think I'd prefer to avoid normalization & have the function focus on being good at a narrow range of things.

nmayorov replied:
I meant before the two previous checks (involving `tol`). To me it looks natural to check that `start` and `end` belongs to a unit sphere first.

> I think I'd prefer to avoid normalization & have the function focus on being good at a narrow range of things.

I think there is no single best solution here --- checking and normalization are two viable options. I'd prefer normalization with an explanation in the docstring. There is also an option to provide a flag like `normalize=False` --- either a user guarantees that he normalized the input or asks us to do that.

tylerjereddy replied:
I did the `np.linalg.norm` thing but I may still need convincing that it is worth reordering the exception handling.

tylerjereddy replied:
I reordered the exception handling as requested here, catching unit sphere violations before the two tolerance-related checks---so marking as resolved.

nmayorov commented on scipy/spatial/_geometric_slerp.py:
```diff
@@ -0,0 +1,206 @@
+    if np.array_equal(start, end):
+        raise ValueError("Start and end coordinates cannot be the same")
+
+    if start.size == 1 or end.size == 1:
```
Perhaps combine this check with the previous one? The error message can be made more straightforward, like "Start and end must have at least 2 elements" (something better than that though).

tylerjereddy replied:
I've fused the two size-related checks into a single check, as requested here. Some unit tests were adjusted for the new error message accordingly.

nmayorov commented on scipy/spatial/_geometric_slerp.py:
```diff
@@ -0,0 +1,206 @@
+                raise ValueError("start and end are not"
+                                 " on a unit n-sphere")
+
+    t = np.asanyarray(t).astype(np.float64)
```
Any benefits from using `asanyarray`? I'd just write `np.asarray(t, dtype=float)`, I think it covers all reasonable usages well.

tylerjereddy replied:
Just a habit from working on NumPy--is there any reason to block the subclass passthrough in the design?

tylerjereddy replied:
I guess you mean the `astype` is coercing anyway?

nmayorov replied:
Well, honestly, if you are deeply familiar with numpy then I can't advice you anything here. My knowledge of numpy is not above a user level.

From a practical experience, I don't think anyone had issues with `np.asarray` used in scipy in many many places, i.e. I just thought that it is a sufficient and simple policy to stick to.

tylerjereddy replied:
Alright, I just did what you suggest, with `np.float64` instead of `float`, but I think that works just as well.

nmayorov commented on scipy/spatial/_geometric_slerp.py:
```diff
@@ -0,0 +1,206 @@
+    >>> plt.show()
+    """
+
+    start = np.asanyarray(start)
```
Same as below --- `np.asarray(start, dtype=float)`?

tylerjereddy replied:
I've done this with `float64` now.

nmayorov commented on scipy/spatial/_geometric_slerp.py:
```diff
@@ -0,0 +1,206 @@
+
+    Returns
+    -------
+    result : (t.size, D)
```
Technically, it might be 1-dimensional if `t` is float. Not sure if it is worth explaining it.

tylerjereddy replied:
Ok, I've expanded the description of the return value to explain this.

nmayorov commented on scipy/spatial/_slerp.pyx:
```diff
@@ -0,0 +1,29 @@
+   cdef:
+      double omega = acos(np.dot(start, end))
+      double sin_omega = sin(omega)
+      int ndims = start.shape[0]
```
`n_pts, n_dims` for similarity? `n_points, n_dimensions` is a good alternative.

pmla commented on scipy/spatial/_slerp.pyx:
```diff
@@ -0,0 +1,29 @@
+                       double[:] t):
+   cdef:
+      double omega = acos(np.dot(start, end))
+      double sin_omega = sin(omega)
```
You can get to `sin_omega` without any trigonometry and better stability using
```
sin_omega = np.linalg.norm(np.cross(P, Q))
```

pmla commented on scipy/spatial/_geometric_slerp.py:
```diff
@@ -0,0 +1,206 @@
+    tol: float
+        The absolute tolerance for determining if the start and end
+        coordinates are antipodes. Antipode detection will
+        raise an exception because of the ambiguity in the path.
```
> Antipode detection will raise an exception because of the ambiguity in the path

Do you want to move this to a `Raises` section?  Just a suggestion.

tylerjereddy replied:
Ok, I've done as you requested here & transplanted to a `Raises` section.

pmla commented:
I think that the standard Slerp definition has poor numerical properties.
If you create an orthogonal basis you should get better numerical stability.  Here is a basic example:

```
def slerp(start, end, t):

    # calculate the angle between `start` and `end`
    s = np.linalg.norm(np.cross(start, end))
    c = np.dot(start, end)
    omega = np.arctan2(s, c)

    # create an orthogonal basis using QR decomposition
    basis = np.vstack([start, end])
    Q = np.linalg.qr(basis.T)[0].T
    signs = np.sign(np.einsum('ij,ij->i', basis, Q))
    start, end = Q * np.array([signs]).T

    # interpolate
    s = np.sin(t * omega)
    c = np.cos(t * omega)
    return start * c[:, np.newaxis] + end * s[:, np.newaxis]
```

nmayorov commented on scipy/spatial/_slerp.pyx:
```diff
@@ -0,0 +1,29 @@
+                       double[:] t):
+   cdef:
+      double omega = acos(np.dot(start, end))
+      double sin_omega = sin(omega)
```
The problem is that `np.cross` works only with 2d and 3d vectors. Do you know how to implement arbitrary dimension generalization of the cross product?

nmayorov replied:
Well, as I quickly figured it is reduced to determinants computation. I'm not sure if it is worth it in this case.

pmla replied:
Oh, I hadn't thought of that. `sqrt(1 - np.dot(start, end)**2)` will achieve the same result without any trig. It might not be worth it anyway.

tylerjereddy replied:
sounds like there isn't a consensus this should be changed, so I'll mark as resolved

tylerjereddy commented:
I've pushed some adjustments to address some (but not all) reviewer comments. Rebased on latest master while I was at it.

tylerjereddy commented:
> I think it also might have better numerical stability as omega tends towards 180 degrees

If that's true, a unit test demonstrating a failure for the current implementation might be helpful.

pmla commented:
I have done some testing.  I have used the deviation from unit norm as the error measure.  Here are the results:
```
       angle         stable        classic
=============================================
pi - 1.000000e-01  1.110223e-16  4.773959e-15
pi - 1.000000e-02  1.110223e-16  1.334488e-13
pi - 1.000000e-03  1.110223e-16  7.992274e-12
pi - 1.000000e-04  1.110223e-16  2.623915e-09
pi - 1.000000e-05  1.110223e-16  4.133418e-08
pi - 1.000000e-06  1.110223e-16  4.443573e-05
pi - 1.000000e-07  1.110223e-16  3.997795e-04
pi - 1.000000e-08  1.110223e-16  8.164592e+07
pi - 1.000000e-09  1.110223e-16  8.164593e+06
pi - 1.000000e-10  1.110223e-16  8.164593e+05
```

I used this script to generate the results:
```
import numpy as np


def slerp(start, end, t):
    omega = np.arccos(np.dot(start, end))
    sin_omega = np.sin(omega)

    a = np.sin((1 - t) * omega) / sin_omega
    b = np.sin(t * omega) / sin_omega
    return a[:, None] * start + b[:, None] * end


def stable_slerp(start, end, t):

    # calculate the angle between `start` and `end`
    s = np.prod(np.linalg.svd([start, end])[1])
    c = np.dot(start, end)
    omega = np.arctan2(s, c)

    # create an orthogonal basis using QR decomposition
    basis = np.vstack([start, end])
    Q = np.linalg.qr(basis.T)[0].T
    signs = np.sign(np.einsum('ij,ij->i', basis, Q))
    start, end = Q * np.array([signs]).T

    # interpolate
    s = np.sin(t * omega)
    c = np.cos(t * omega)
    return start * c[:, np.newaxis] + end * s[:, np.newaxis]


ts = np.linspace(0, 1, 100)

print("       angle         stable        classic")
print("=============================================")
for k in np.logspace(-10, -1, 10)[::-1]:
    P = np.array([1, 0, 0, 0])
    Q = np.array([np.cos(np.pi-k), np.sin(np.pi-k), 0, 0])

    norms1 = np.linalg.norm(stable_slerp(P, Q, ts), axis=1)
    norms2 = np.linalg.norm(slerp(P, Q, ts), axis=1)
    error1 = np.max(np.abs(norms1 - 1))
    error2 = np.max(np.abs(norms2 - 1))
    print("pi - %e  %e  %e" % (k, error1, error2))
```

nmayorov commented on scipy/spatial/_geometric_slerp.py:
```diff
@@ -0,0 +1,209 @@
+    # performing the check for 1-sphere (circle) and up
+    if start.size > 1:
+        for coord in [start, end]:
+            if not np.allclose(np.square(coord).sum(), 1.0, rtol=0, atol=tol):
```
I think there is no reason not to use `np.linalg.norm` --- it does a stable computation and it doesn't create a bottleneck.

As for the tolerance --- a proper normalized array must have norm equal to 1.0 within machine epsilon. So I would consider using `atol=0, rtol=10 * EPS` (for example).

tylerjereddy replied:
I'm using `np.linalg.norm` now---the tolerances were adjusted to allow the tests to continue passing; `rtol` couldn't be as close to double precision limit as I might like, but maybe good enough

tylerjereddy commented:
I've revised substantially based on reviewer feedback---see the most recent commit for details. In short, I removed the Cython implementation and used the "stable" Python implementation demonstrated above. 

That said, there were still some accuracy tradeoffs for some tests, but I ultimately went with removal of the Cython code based on the reviewer feedback.

Probably still some more clean-ups needed, but perhaps a bit less controversial now. I can deal with Cythonizing if/when that is needed.

pmla commented on scipy/spatial/_geometric_slerp.py:
```diff
@@ -0,0 +1,224 @@
+    s = np.sin(t * omega)
+    c = np.cos(t * omega)
+    return start * c[:, np.newaxis] + end * s[:, np.newaxis]
+
```
I have given my code some more thought. The SVD is not necessary since we do a QR decomposition. Here is the simplified code:
```
def _geometric_slerp(start, end, t):
    # create an orthogonal basis using QR decomposition
    basis = np.vstack([start, end])
    Q, R = np.linalg.qr(basis.T)
    signs = np.sign(np.diag(R))
    Q = Q.T * signs.T[:, np.newaxis]
    R = R.T * signs.T[:, np.newaxis]
    start, end = Q

    # calculate the angle between `start` and `end`
    c = np.dot(start, end)
    s = np.linalg.det(R)
    omega = np.arctan2(s, c)

    # interpolate
    s = np.sin(t * omega)
    c = np.cos(t * omega)
    return start * c[:, np.newaxis] + end * s[:, np.newaxis]
```
The relationship between the SVD and the QR is described here:
https://en.wikipedia.org/wiki/QR_decomposition#Connection_to_a_determinant_or_a_product_of_eigenvalues

tylerjereddy replied:
This doesn't pass the test suite at the moment---in particular, `TestGeometricSlerp.test_include_ends` fails, suggesting that `start` or `end` isn't being included anymore.

Looking a little more closely, it seems that `end` is excluded here now.

tylerjereddy replied:
In the meanwhile, I've added `asv` benchmarks, so that if you correct that code & stand by the claim that it is more efficient we can test the performance.

At the moment, `asv run -b 'GeometricSlerp*'`:

```
[100.00%] ··· ....GeometricSlerpBench.time_geometric_slerp_3d                 ok
[100.00%] ··· ============ =========
               num_points
              ------------ ---------
                   10       321±3μs
                  1000      381±2μs
                 10000      819±4μs
              ============ =========
```

pmla replied:
Thanks. There was indeed a bug in the code. Here is the fixed code:
```
def _geometric_slerp(start, end, t):
    # create an orthogonal basis using QR decomposition
    basis = np.vstack([start, end])
    Q, R = np.linalg.qr(basis.T)
    signs = np.sign(np.diag(R))
    Q = Q.T * signs.T[:, np.newaxis]
    R = R.T * signs.T[:, np.newaxis]

    # calculate the angle between `start` and `end`
    c = np.dot(start, end)
    s = np.linalg.det(R)
    omega = np.arctan2(s, c)

    # interpolate
    start, end = Q
    s = np.sin(t * omega)
    c = np.cos(t * omega)
    return start * c[:, np.newaxis] + end * s[:, np.newaxis]
```
Benchmarks on my laptop:
```
============ ============ ============
 num_points    with svd     qr only
------------ ------------ ------------
     10       39.0±0.3ms   31.2±0.4ms
    1000       46.7±1ms    39.4±0.2ms
   10000       114±1ms     95.9±0.3ms
============ ============ ============
```
It is faster but not by much, so pick the version you prefer.

I wrapped the benchmark in a for loop to get millisecond rather than nanosecond times:
```
    def time_geometric_slerp_3d(self, num_points):
        # time geometric_slerp() for 3D interpolation
        for i in range(100):
            geometric_slerp(start=self.start,
                            end=self.end,
                            t=self.t)
```

tylerjereddy replied:
Ok, I pulled in your most recent implementation, though conventional benchmarks show that both approaches are effectively the same within the default performance threshold.

tylerjereddy commented on scipy/spatial/tests/test_slerp.py:
```diff
@@ -0,0 +1,345 @@
+from __future__ import division, absolute_import, print_function
+
+import math
```
Not sure, but easier for me to just make the change you request, which I've done now.

tylerjereddy commented:
I note that there are still some outstanding comments about the order that exceptions are handled, and a few comments about documentation.

larsoner commented:
@pmla @nmayorov happy with this latest version?

pmla commented:
I'm happy.

nmayorov commented on scipy/spatial/_geometric_slerp.py:
```diff
@@ -0,0 +1,226 @@
+    # diameter of 2 within tolerance means antipodes, which is a problem
+    # for all unit n-spheres (even the 0-sphere would have an ambiguous path)
+    if np.allclose(coord_dist, 2.0, rtol=0, atol=tol):
+        raise ValueError("start and end are antipodes"
```
I don't know --- should it be only a warning about possible ambiguity? Can `tol` be determined automatically? (Possibly much lower than `1e-7`, on the order of `1e-15` I suppose.)

tylerjereddy replied:
No, it is not appropriate to emit a warning here---antipodes are **degenerate** input, the result isn't even on the unit sphere---if I disable the guard:

```python
import numpy as np
import scipy
from scipy.spatial import geometric_slerp
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import proj3d

path = geometric_slerp(start=[0, 0, 1],
                         end=[0, 0, -1],
                         t=np.linspace(0, 1, 50))

fig = plt.figure()
ax = fig.add_subplot(111, projection='3d')

# plot the unit sphere for reference ----
u = np.linspace(0, 2 * np.pi, 100)
v = np.linspace(0, np.pi, 100)
x = np.outer(np.cos(u), np.sin(v))
y = np.outer(np.sin(u), np.sin(v))
z = np.outer(np.ones(np.size(u)), np.cos(v))
ax.plot_surface(x, y, z, color='y', alpha=0.1)
# end plot unit sphere for reference ----

ax.plot_surface(x, y, z, color='y', alpha=0.1)
ax.scatter(path[..., 0],
           path[..., 1],
           path[..., 2])
ax.azim = 50
ax.elev = 80

# antipodes are degenerate input---the path is not
# even on the unit sphere:
print(np.linalg.norm(path, axis=1))

fig.savefig('antipodal_paths.png',
            dpi=300)
```

```
[1.         0.99794539 0.99179001 0.98155916 0.96729486 0.94905575
 0.92691676 0.90096887 0.8713187  0.8380881  0.80141362 0.76144596
 0.71834935 0.67230089 0.6234898  0.57211666 0.51839257 0.46253829
 0.40478334 0.34536505 0.28452759 0.22252093 0.1595999  0.09602303
 0.03205158 0.03205158 0.09602303 0.1595999  0.22252093 0.28452759
 0.34536505 0.40478334 0.46253829 0.51839257 0.57211666 0.6234898
 0.67230089 0.71834935 0.76144596 0.80141362 0.8380881  0.8713187
 0.90096887 0.92691676 0.94905575 0.96729486 0.98155916 0.99179001
 0.99794539 1.        ]
```

![antipodal_paths](https://user-images.githubusercontent.com/7903078/69012549-4d6a8f80-0934-11ea-8463-87d2600af6e9.png)

Now, if we move away from antipodes with `start=[0, np.sqrt(2)/2, np.sqrt(2)/2]`:

```
[1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.
 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.
 1. 1.]
```

![antipodal_paths](https://user-images.githubusercontent.com/7903078/69012593-b5b97100-0934-11ea-9ccc-d41792639bc7.png)

If we were to coerce Peter's algorithm to produce something less pathological for antipodes, I still think it is a bad idea---clearly defining what is degenerate input is actually helpful for computational geometry algorithms.

If an end user wants a bunch of random great circle arcs between poles they are free to opt in **explicitly** by choosing random points on a circle and then use `geometric_slerp` to define non-ambiguous paths from the poles to the equators & then fusing those data sets.

pmla replied:
There was a silly bug which causes this pathological output. See my comment above.

With this in mind, I think that raising an exception for degenerate input is too strict. I would go with a warning at most.

tylerjereddy replied:
I've changed it to a `UserWarning`, though I still disagree here. There's an infinite set of valid return values/paths now. Here is a plot of one of them, which is at least on the unit sphere now:

![antipodal_paths](https://user-images.githubusercontent.com/7903078/74704242-d65fc400-51cc-11ea-84ac-d55ecf78a03a.png)

Anyway, I'm not going to win this battle, so I've applied the patch & will mark as resolved.

nmayorov commented on scipy/spatial/_geometric_slerp.py:
```diff
@@ -0,0 +1,226 @@
+
+    # a diameter > 2 within tol violates requirement for input to be on
+    # unit n-sphere
+    if coord_dist > (2.0 + tol):
```
Looks like checks on line 204 handles this violation. Should it be removed?

tylerjereddy replied:
Alright, I've removed the redundant error handling here.

tylerjereddy commented:
I'll bump the milestone to `1.5.0`, but let's try to wrap up the discussion on error/warning handling soon. Possibly a third person can just push in whatever changes they prefer for the remaining issues there & I don't mind.

tylerjereddy commented on scipy/spatial/_geometric_slerp.py:
```diff
@@ -0,0 +1,206 @@
+        Single D-dimensional input coordinate in a 1-D array-like
+        object.
+        Starting coordinate for interpolation will be converted
+        to float64.
```
Ok, I've removed the indications that `start` and `end`  are converted to `float64` in the docstring.

tylerjereddy commented:
@nmayorov @WarrenWeckesser Are you a bit happier here now?

pmla commented on scipy/spatial/_geometric_slerp.py:
```diff
@@ -0,0 +1,218 @@
+    # create an orthogonal basis using QR decomposition
+    basis = np.vstack([start, end])
+    Q, R = np.linalg.qr(basis.T)
+    signs = np.sign(np.diag(R))
```
In the case of degenerate input, this sign computation destroys the relationship between the `basis` variable and its QR decomposition. Sorry that I did not think of this.

If we replace it by:
`    signs = 2 * (np.diag(R) >= 0) - 1`
(which effectively disallows zero signs) the QR relationship is preserved. This also means that degenerate input no longer goes through the origin and looks like a normal slerp.

pmla replied:
I often have a need to correct the signs of a QR decomposition. From wikipedia:
>...the factorization is unique if we require the diagonal elements of R to be positive.

@tylerjereddy I have often thought that this would be worth fixing upstream, i.e. in numpy. Do you think they would be receptive to that?

tylerjereddy replied:
Probably best to bring that up with the NumPy team, I've applied your patch here & adjusted tests accordingly.

pmla replied:
I made an issue: https://github.com/numpy/numpy/issues/15628

tylerjereddy commented:
Probably doc build issues because of the exception->warning transition. I'll try to fix that tomorrow.

tylerjereddy commented:
Ok, CI is all green & the docstring render looks pretty solid to me.

Not going to assume that all points of contention are fully resolved yet though. May also want to hit the mailing list if/when it gets close I suppose.

nmayorov commented on scipy/spatial/_geometric_slerp.py:
```diff
@@ -0,0 +1,226 @@
+    # for points that violate equation for n-sphere
+    for coord in [start, end]:
+        if not np.allclose(np.linalg.norm(coord), 1.0,
+                           rtol=np.finfo(np.float32).eps,
```
Did you actually mean float32 tolerance here?

tylerjereddy replied:
I've adjusted to `1e-9`, which is a bit better.

nmayorov commented on scipy/spatial/_geometric_slerp.py:
```diff
@@ -0,0 +1,226 @@
+    if np.allclose(coord_dist, 2.0, rtol=0, atol=tol):
+        warnings.warn("start and end are antipodes"
+                      " using the specified tolerance or they"
+                      " are not on a unit n-sphere; this may"
```
After checking start and end norms, is it possible that they are not on a unit sphere? That is should it be mentioned here in the warning?

tylerjereddy replied:
Right, the warning has been simplified to mention antipodes only.

nmayorov commented on scipy/spatial/_geometric_slerp.py:
```diff
@@ -0,0 +1,226 @@
+        raise ValueError("interpolation parameter must be in [0, 1]")
+
+    if np.ndim(t) == 0:
+        return _geometric_slerp(start.astype(np.float64),
```
You have cast `start` and `end` to float64 already.

tylerjereddy replied:
I've removed the 4 redundant casts, thanks.

nmayorov commented on scipy/spatial/_geometric_slerp.py:
```diff
@@ -0,0 +1,226 @@
+    if t.min() < 0 or t.max() > 1:
+        raise ValueError("interpolation parameter must be in [0, 1]")
+
+    if np.ndim(t) == 0:
```
Might use `t.ndim` for consistency with other code.

tylerjereddy replied:
Done

nmayorov commented:
@tylerjereddy 

I'm very sorry, I've disappeared from the discussion. I've left several minor comments now. Other than that, +1 to merge from me.

nmayorov commented:
Nothing prevents merging, correct?

tylerjereddy commented:
I think it is in good shape. You or @pmla can merge if happy I think.

Our CI is a little wonky at the moment if we rebase first.

WarrenWeckesser commented:
Thanks for your patience and persistence, @tylerjereddy.  Sorry for not getting back to this sooner.

A quick question: why can't the function handle the case `start == end`?  This should act the same way `linspace` does, and just repeat the point.

E.g.

```
In [24]: np.linspace(1, 1, 8)                                                                                             
Out[24]: array([1., 1., 1., 1., 1., 1., 1., 1.])
```
I expect `geometric_slerp([0, 1], [0, 1], [0, 0.5, 1])` to return `[[0, 1], [0, 1], [0, 1]]`, but it generates an error:

```
In [25]: geometric_slerp([0, 1], [0, 1], [0, 0.5, 1])                                                                     
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-25-ff79027ea298> in <module>
----> 1 geometric_slerp([0, 1], [0, 1], [0, 0.5, 1])

~/mc37scipy/lib/python3.7/site-packages/scipy-1.5.0.dev0+881ab8f-py3.7-macosx-10.7-x86_64.egg/scipy/spatial/_geometric_slerp.py in geometric_slerp(start, end, t, tol)
    184 
    185     if np.array_equal(start, end):
--> 186         raise ValueError("Start and end coordinates cannot be the same")
    187 
    188     # for points that violate equation for n-sphere

ValueError: Start and end coordinates cannot be the same
```

WarrenWeckesser commented on scipy/spatial/tests/test_slerp.py:
```diff
@@ -0,0 +1,378 @@
+                            t=np.linspace(0, 1, 5))
+
+    @pytest.mark.parametrize('k',
+    np.logspace(-10, -1, 10),
```
`pycodestyle` doesn't like this line, but I haven't tried to figure out what the "correct" spacing should be to avoid over-indent or under-indent issues from `pycodestyle`.  In this case, the entire decorator easily fits in one line, so why not just write

    @pytest.mark.parametrize('k', np.logspace(-10, -1, 10))

tylerjereddy commented:
Alright, I fixed all the pep8 nitpicks. I also made the function do what Warren wants for `start==end`, irrespective of my view on this.

Rebased on latest master/force-push as well.
